---
title: "CMDMultiOmics: a multi-omits atlas to investigate the complexity and spectrum
  of human cardiometabolic related disease"
author: "Brendan Gongol, Xueyuan Jiang, Xiaowen Chen, Yang Liu"
# date: "2024-02-09"
output: 
  BiocStyle::html_document:
    toc_float: true
    toc_depth: 3
    cold_folding: show
    number_sections: false
always_allow_html: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



# About

The `CMDMultiOmics` package provides a tool set within a shiny app environment for the integration of proprietary and open source data sets. Currently the `CMDMultiOmics` app houses custom selection of data that are related to a group of cardiometabolic diseases from trascriptomics, proteomics, methylation or metabolomic data types. This shiny application is paired with transcript expression and proteomics processing pipelines for the addition of user selected datasets. Details rgarding these workflows are escribed in this vignette which ends with the deployment of the shiny app.

# Installation and setup

There are three key steps involved in the installation and launch of the `CMDMultiOmics` package as outlined below. <br>

1) Install dependencies <br>
```{r, warning=FALSE, message=FALSE, eval=FALSE}
install.packages(c("plotly", "shinyWidgets", "data.table", "feather", "textclean", "dplyr", "ggplot2", "RColorBrewer", "stringr", "readxl", "R.utils", "umap", "tidyr", "tidyverse", "makeunique", "radiant.data"))

BiocManager::install(c("Biobase", "limma", "DESeq2", "org.Hs.eg.db", "org.Mm.eg.db", "org.Rn.eg.db", "GEOquery", "mia", "clariomdhumantranscriptcluster.db", "hugene10sttranscriptcluster.db", "hugene20sttranscriptcluster.db", "hugene11sttranscriptcluster.db", "hgu133plus2.db", "hwgcod.db", "illuminaHumanv4.db", "hta20transcriptcluster.db", "hgu133a.db", "AnnotationDbi", "rpx", "RforProteomics", "BiocFileCache", "DEP", "NormalyzerDE", "SummarizedExperiment"))
```

2) Install app <br>
```{r, warning=FALSE, message=FALSE, eval=FALSE}
devtools::install_github("brgongol/MultiOmics")
```

3) Execute app deployment <br>
In this step, the under must specify the path by editing the "path to database/MultiOmicsAnalysis" line of code where the database linked to the app will be located. The `makeDirectory` function then builds the specified directory containing the file structures needed to execute the remainder of this vignette and launch the application.   
```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(MultiOmics)
homedir <- "path to database/MultiOmicsAnalysis"
makeDirectory(homedir)
setwd(homedir)
```
One the `makeDirectory` function has been executed, open the `Scripts` directory and load and execute the "WorkflowScript.R" file to install the database and load the shiny app.

file structure and directory organization <br>

**AppData** <br>
**ArraymetaData** <br>
**DirectionCheck** <br>
**ExternalAnalyzed** <br>
**GEOcache** <br>
**Metabolomics** <br>
**Methylation** <br>
**OverviewFiles** <br>
**ProcessFiles** <br>
**Proteomic_1** <br>
**Proteomic_2** <br>
**Proteomic_3** <br>
**RawQC** <br>
**RunInfo** <br>
**Scripts** <br>


Set up environment
```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(MultiOmics); library(data.table)
homedir <- "path to database/MultiOmicsAnalysis"
setwd(homedir)
```

Obtain platform information
```{r, warning=FALSE, message=FALSE, eval=FALSE}
PlatAnnotInfo <- PlatformAnnotationLoad(PlatInfo=fread(file.path(homedir, "OverviewFiles", "GEOPlatformInfo.csv"), header = TRUE))
fwrite(PlatAnnotInfo, file.path(homedir, "OverviewFiles", "GEOPlatformAnnotation.txt"), row.names = FALSE, quote = FALSE, sep = "\t")
PlatAnnotInfo <- fread(file.path(homedir, "OverviewFiles", "GEOPlatformAnnotation.txt"))
```

Load overview file
```{r, warning=FALSE, message=FALSE, eval=FALSE}
overview <- fread(file.path(homedir, "OverviewFiles", "GEODataOverview3.csv"), header = TRUE)
overview <- overview[GEO2R == TRUE,]
#### check for duplicates assay names ####
overview$AssayNames <-gsub(" ", "", paste(overview$ID, overview$FCColumnNames, overview$Tissue, overview$Disease, sep = "_"))
overview[duplicated(AssayNames),][!AssayNames == "__",] # check should return nothing
```

Download data from GEO
```{r, warning=FALSE, message=FALSE, eval=FALSE}
Compiled <- GEOCompile(DS=overview$ID,
                       gpl=overview$GPLNumber,
                       gsm=overview$ComparisonVector,
                       namestr=overview$FCColumnNames,
                       nameraw=overview$RawColumnNames,
                       PlatAnnotInfo = PlatAnnotInfo,
                       destdir = file.path(homedir, "GEOcache"),
                       filename = NULL,
                       writeDB=TRUE,
                       writeRaw=overview$DownloadRawData,
                       GenerateMetaData=overview$DownloadMetaData,
                       MetaDataPath = file.path(homedir, "ArrayMetaData"),
                       writeMetaData=TRUE,
                       DBPath=file.path(homedir, "AppData"),
                       Technology = overview$Technology,
                       renameRaw = FALSE,
                       subsetRaw = FALSE)
```

Double check that the fold change calculations were performed correctly
```{r, warning=FALSE, message=FALSE, eval=FALSE}
checkFile <- GEO2RDirectionCheck(DBPath=file.path(homedir, "AppData"),
                                 DS=overview$ID,
                                 namestr=overview$FCColumnNames,
                                 gsm=overview$ComparisonVector,
                                 Technology = overview$Technology,
                                 GraphPath = file.path(homedir, "DirectionCheck"),
                                 subsetRaw = FALSE,
                                 writeRaw=overview$DownloadRawData,
                                 RawQCPath = file.path(homedir, "RawQC"))
checkFile[[1]][correlation < 0,]
fwrite(checkFile[[1]], file.path("./OverviewFiles/ArrayCheck.xls"), row.names = FALSE, quote = FALSE, sep = "\t")
```

Externally analyzed RNAseq data setup
```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(org.Hs.eg.db); library(org.Mm.eg.db); library(org.Rn.eg.db)
ExternalDataHarmonize(Fpath = file.path(homedir, "ExternalAnalyzed"),
                      OutPath = file.path(homedir, "AppData"))
```

Create or update Summarized experiment objects that integrate all data sets: DEG data only <br>
Checks for comparisons that are already in the database <br>
Duplicated comparisons are not deleted from "AppData" <br>
```{r, warning=FALSE, message=FALSE, eval=FALSE}
(DESE <- DESEGenerate(DEGDatapath=file.path(homedir, "AppData"), SEPath = file.path(homedir, "ProcessFiles", "SumarizedExp_DB.rds") ))
(DESE <- readRDS(file.path(homedir, "ProcessFiles", "SumarizedExp_DB.rds")))
```

Download GTF file
```{r, warning=FALSE, message=FALSE, eval=FALSE}
options(timeout=24000); getOption('timeout')
download.file("https://ftp.ensembl.org/pub/current_gtf/homo_sapiens/Homo_sapiens.GRCh38.111.chr.gtf.gz",
              destfile = file.path(homedir, "OverviewFiles", "GTFHuman.gtf.gz"), quiet = FALSE)
download.file("https://ftp.ensembl.org/pub/current_gtf/mus_musculus/Mus_musculus.GRCm39.111.chr.gtf.gz",
              destfile = file.path(homedir, "OverviewFiles", "GTFMouse.gtf.gz"), quiet = FALSE)
```

Create or update Raw data files <br>
Issue of duplicated genes here probably because of multiple probes measuring the same gene. dealt with this by averaging
```{r, warning=FALSE, message=FALSE, eval=FALSE}
RawDataCompile(Fpath = file.path(homedir, "AppData"),
               outPath = file.path("./ProcessFiles/RawData.txt"),
               StartAt = 1,
               sleep = 60, # DataType = "Array",
               GTFHumanFpath = file.path(homedir, "OverviewFiles", "GTFHuman.gtf.gz"),
               GTFMouseFpath = file.path(homedir, "OverviewFiles", "GTFMouse.gtf.gz") )
RawArrayComplete <- fread(file.path("./ProcessFiles/RawData.txt"))
```

Create meta data for Raw data file
```{r, warning=FALSE, message=FALSE, eval=FALSE}
metaDF <- MetDataCompile(RNAseqFilePath= file.path("RunInfo"), ArrayFilePath= file.path("ArrayMetaData"), overview=overview)
#### Save meta data data frame ####
metaDF$rownames <- rownames(metaDF)
fwrite(metaDF, file.path(homedir, "OverviewFiles", "metaData.xls"), row.names = FALSE, quote = FALSE, sep = "\t")
metaDF <- as.data.frame(fread(file.path(homedir, "OverviewFiles", "metaData.xls")))
rownames(metaDF) <- metaDF$rownames
metaDF$rownames <- NULL
```

Create raw data summarizedExperiment object
```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(SummarizedExperiment)
RawSEList <- GenerateRawSE(df = metaDF, ArrayDT=RawArrayComplete, GTFFpath = file.path(homedir, "OverviewFiles", "GTFHuman.gtf.gz"), overview=overview)
names(RawSEList)
saveRDS(RawSEList[["RawSE"]], file=file.path(homedir, "ProcessFiles", "SumarizedExp_RawDB.rds"))
RawSE <- readRDS(file.path(homedir, "ProcessFiles", "SumarizedExp_RawDB.rds"))
assay(RawSEList[["RPKMSE"]])[1:5,1:5]; head(rowData(RawSEList[["RPKMSE"]])); head(colData(RawSEList[["RPKMSE"]]))
saveRDS(RawSEList[["RPKMSE"]], file=file.path(homedir, "ProcessFiles", "expression_norm.v2.RDS"))
expression_norm <- readRDS(file.path(homedir, "ProcessFiles", "expression_norm.v2.RDS"))
assay(expression_norm)[1:5,1:5]; head(rowData(expression_norm)); head(colData(expression_norm))
```

Update overview file with Externally processed data
```{r, warning=FALSE, message=FALSE, eval=FALSE}
overview <- fread(file.path(homedir, "OverviewFiles", "GEODataOverview3.csv"), header = TRUE)
overview <- overview[!ID == "",]
IntDT <- data.table(ID = gsub("_.+", "", names(assays(DESE))), FCColumnNames = gsub("^.+?_", "", names(assays(DESE))))
IntDT <- IntDT[!(IntDT$ID %in% overview$ID),] # [grepl("PID[0-9]+", ID),]
IntDT[, `:=`(DataType = "BulkExpressionProfile",	Technology = "RNAseq", DownloadMetaData = FALSE, DownloadRawData = FALSE, GEO2R = FALSE)]
overview <- rbind(overview, IntDT, fill = TRUE)
#### overwrite previous file ####
fwrite(overview, file.path(homedir, "OverviewFiles", "GEODataOverview3.xls"), row.names = FALSE, quote = FALSE, sep = "\t")
#### Update date and other columns by hand ####
```

# Proteomic data workflow

Load overview file
```{r, warning=FALSE, message=FALSE, eval=FALSE}
overview <- fread(file.path(homedir, "OverviewFiles", "GEODataOverview3.csv"), header = TRUE)
overviewpProteomics <- overview[DataType == "Proteomic",]
```

Download proteomics data
```{r, warning=FALSE, message=FALSE, eval=FALSE}
# devtools::install_version("dbplyr", version = "2.3.4")
# ProteomicsDataDownload(path = file.path(homedir, "Proteomic_1"), DS = overviewpProteomics$ID)
```

Loop through files and format data
```{r, warning=FALSE, message=FALSE, eval=FALSE}
# MZList <- FormatMaxQuant(path = file.path(homedir, "Proteomic_1"))
# names(MZList)
```

Save formatted proteomics data
```{r, warning=FALSE, message=FALSE, eval=FALSE}
# proteomicMZSave(MZList=MZList, path = file.path(homedir, "Proteomic_2"))
```

Create experimental design object
```{r, warning=FALSE, message=FALSE, eval=FALSE}
DesignDT <- DesignMatrixFromNames(Fpath = file.path(homedir, "Proteomic_2"))
fwrite(DesignDT, file.path(homedir, "OverviewFiles", "DesignMatrix.xls"), row.names = FALSE, quote = FALSE, sep = "\t")
```

Load into summarized experiment object
```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(SummarizedExperiment)
tissueSplitList <- ProtSELoad(DesignDT=DesignDT, Fpath=file.path(homedir, "Proteomic_2"))
assay(tissueSplitList[[2]])
rowData(tissueSplitList[[2]])
```

Perform overall protein abundance assessments <br>
compile data together to assess protein expression level
```{r, warning=FALSE, message=FALSE, eval=FALSE}
TotMel <- RowDataCompile(tissueSplitList=tissueSplitList)
TotMel
```

Filter for proteins that are identified in all replicates of at least one condition
```{r, warning=FALSE, message=FALSE, eval=FALSE}
dataFiltList <- DataFilter(dataSeList=tissueSplitList, thr = 0)
assay(dataFiltList[[3]])
rowData(dataFiltList[[3]])
```

Perform N-peptides per protein cutoff
```{r, warning=FALSE, message=FALSE, eval=FALSE}
PepCutOffList <- NPeptideThreshold(dataFiltList=dataFiltList, Npeptides = 2)
PepCutOffList[[2]]                  # return the percent of records remaining
dataPepCutOff <- PepCutOffList[[1]] # return the data
colnames(assay(dataPepCutOff[[1]]))
```

Perform normalization
```{r, warning=FALSE, message=FALSE, eval=FALSE}
MultiNormalizeList <- MultiNormalization(dataPepCutOff=dataPepCutOff)
names(MultiNormalizeList)
```

Create normalization density plots
```{r, warning=FALSE, message=FALSE, eval=FALSE}
densityPlotList <- densityPlotFromList(MultiNormalizeList)
names(densityPlotList)
densityPlotList[["mean"]]
densityPlotList[["median"]]
densityPlotList[["vsn"]]
densityPlotList[["DEPvsn"]]
densityPlotList[["loess"]]
densityPlotList[["rlr"]]
densityPlotList[["smad"]]
```

Select normalization method
```{r, warning=FALSE, message=FALSE, eval=FALSE}
NormalizedSE <- MultiNormalizeList[["median"]]
assay(NormalizedSE[[3]])
rowData(NormalizedSE[[3]])
```

Impute missing data
```{r, warning=FALSE, message=FALSE, eval=FALSE}
#### Determine if any data sets have missing values that need to be imputed ####
(Missing <- DetermineMising(data=NormalizedSE))
#### Impute missing values ####
set.seed(1)
NormImpAll <- DataImpute(dataFiltList = NormalizedSE, type = "MinProb")
```

Remove samples if there are less than 50 samples in them
```{r, warning=FALSE, message=FALSE, eval=FALSE}
VarRMThreshList <- LowSampleCountRmove(VarRMList=NormImpAll, cut = 50)
```

Calculate Fold changes on imputed, thresholded, and normalized data
```{r, warning=FALSE, message=FALSE, eval=FALSE}
FCcut <- log2(1); Pcut <- 0.05; sigCol = "BHCorrection" # "p.val" "p.adj" "BHCorrection"
Comparisons <- fread(file.path(homedir, "OverviewFiles", "ProteomicComparisons.xls"), header = TRUE)
compList2 <- Comparisons$Comparison
names(compList2) <- Comparisons$dataset
compList2 <- compList2[names(compList2) %in% names(VarRMThreshList)]
dataDiffNorm <- DEAnalysis(DataList = VarRMThreshList, type = "manual", ComparisonList=compList2)
dataDiffNorm
```

Save data to database
```{r, warning=FALSE, message=FALSE, eval=FALSE}
SaveToProteomicDB(SEList=dataDiffNorm, Path=file.path(homedir, "Proteomic_3"))
```

Setup Protein name DB
```{r, warning=FALSE, message=FALSE, eval=FALSE}
Proteins <- ProteomicProteinName(fPath = "./Proteomic_3")
saveRDS(Proteins, file.path(homedir, "OverviewFiles", "ProteomicProteins.RDS"))
```

# Load app

```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(shiny)
AppSetup(homedir)
runApp("./Scripts")
```










